### PR TITLE
mfa: per-session U2F challenge for web SSH

### PR DESCRIPTION
**Pull request by awly**
Check whether MFA is required for the current session and send a
challenge over the websocket.

`client.IssueUserCertsWithMFA` had to be modified to inject proxy's
cached user certs and websocket-based U2F prompt.

### CODE REVIEW

**Review by a-palchikov — APPROVED** [review]

File: lib/client/client.go [hunk 1]
```diff
@@ -254,12 +268,18 @@ func (proxy *ProxyClient) reissueUserCerts(ctx context.Context, params ReissuePa
 	return key, nil
 }
 
+type PromptMFAChallengeHandler func(ctx context.Context, proxyAddr string, c *proto.MFAAuthenticateChallenge, promptDevicePrefix string) (*proto.MFAAuthenticateResponse, error)
```
Comment (r0mant) [inline lib/client/client.go:271]: Nit: godoc.

File: lib/client/client.go [hunk 2]
```diff
@@ -254,12 +268,18 @@ func (proxy *ProxyClient) reissueUserCerts(ctx context.Context, params ReissuePa
...
-	localAgent := proxy.teleportClient.LocalAgent()
-	key, err := localAgent.GetKey(WithKubeCerts(params.RouteToCluster))
-	if err != nil {
-		return nil, trace.Wrap(err)
+func (proxy *ProxyClient) IssueUserCertsWithMFA(ctx context.Context, params ReissueParams, promptMFAChallenge PromptMFAChallengeHandler) (*Key, error) {
+	key := params.ExistingCreds
+	if key == nil {
+		localAgent := proxy.teleportClient.LocalAgent()
+		var err error
+		key, err = localAgent.GetKey(WithKubeCerts(params.RouteToCluster))
```
Comment (r0mant) [inline lib/client/client.go:279]: According to the code below, it's possible that `params.RouteToCluster` will be empty here. Is that gonna be handled properly?

**Review by r0mant — APPROVED** [review]

File: lib/client/client.go (hunk 2 above)
Comment (awly) [reply lib/client/client.go:288]: Added defaulting logic for `params.RouteToCluster`

**Review by russjones — APPROVED** [review]
Bot.

**Pull request merged**